bam_sort.c

```diff
@@ -140,376 +239,721 @@ static void trans_tbl_destroy(trans_tbl_t *tbl) {
+                    __func__, merged_hdr->target_name[i + old_n_targets]);
+            missing++;
         }
-        regfree(&pg_id_search);
+    }
+    if (missing) goto fail;
 
-        // Insert it into our translation map
-        int in_there = 0;
-        khiter_t iter = kh_put(c2c, tbl->pg_trans, ks_release(&match_id), &in_there);
-        char *transformed_id_s = ks_release(&transformed_id);
-        kh_value(tbl->pg_trans,iter) = transformed_id_s;
-        // take matched line and replace ID with transformed_id
-        kstring_t transformed_line = { 0, 0, NULL };
-        if (transformed_equals_match) {
-            kputsn(text+matches[0].rm_so, matches[0].rm_eo-matches[0].rm_so, &transformed_line);
+    free(new_sq_matches);
+    return 0;
+
+ memfail:
+    perror(__func__);
+ fail:
+    free(new_sq_matches);
+    return -1;
+}
+
+/*
+ * Common code for setting up RG and PG record ID tag translation.
+ *
+ * is_rg is true for RG translation, false for PG.
+ * translate is the input bam header
+ * merge is true if tags with the same ID are to be merged.
+ * known_ids is the set of IDs already in the output header.
+ * id_map is the translation map from input header IDs to output header IDs
+ * If override is set, it will be used to replace the existing ID (RG only)
+ *
+ * known_ids and id_map have entries for the new IDs added to them.
+ *
+ * Return value is a linked list of header lines with the translated IDs,
+ * or NULL if something went wrong (probably out of memory).
+ *
+ */
+
+static klist_t(hdrln) * trans_rg_pg(bool is_rg, bam_hdr_t *translate,
+                                    bool merge, khash_t(cset)* known_ids,
+                                    khash_t(c2c)* id_map, char *override) {
+    hdr_match_t matches[2];
+    khiter_t iter;
+    const char *text = translate->text;
+    const char *rec_type = is_rg ? "@RG" : "@PG";
+    klist_t(hdrln) *hdr_lines;
+
+    hdr_lines = kl_init(hdrln);
+
+    // Search through translate's header
+    while (hdr_line_match(text, rec_type, "ID", matches) == 0) {
+        // matches[0] is the whole @RG/PG line; matches[1] is the ID field value
+
+        kstring_t orig_id = { 0, 0, NULL };        // ID in original header
+        kstring_t transformed_id = { 0, 0, NULL }; // ID in output header
+        char *map_value;    // Value to store in id_map
+        bool id_changed;    // Have we changed the ID?
+        bool not_found_in_output; // ID isn't in the output header (yet)
+
+        // Take a copy of the ID as we'll need it for a hash key.
+        if (match_to_ks(text, &matches[1], &orig_id)) goto memfail;
+        
+        // is our matched ID in our output ID set already?
+        iter = kh_get(cset, known_ids, ks_str(&orig_id));
+        not_found_in_output = (iter == kh_end(known_ids));
+
+        if (override) {
+            // Override original ID (RG only)
+#ifdef OVERRIDE_DOES_NOT_MERGE
```

Comment:
This is the question of what to do when the -r option is in use (to force a read group for the input file) but the input file contains more than one read group.   At the moment they all end up with the same RG, which is similar to what it did before.  Defining OVERRIDE_DOES_NOT_MERGE makes it generate unique IDs for each RG instead, with override as  a prefix.  That is a change in behaviour, and also makes the tests fail, so some discussion is needed before we decide to go that way.  Hence it's there but not enabled at the moment.